_posts/2023-02-16-announcing-ballerina-2201.4.0-swan-lake-update-4.md

```diff
@@ -0,0 +1,51 @@
+
+The rest of the significant additions and improvements are described below.
+
+## Runtime updates
+
+Now, you can obtain the strand dump when running the tests in a Ballerina package or a file using the bal test command by sending the `SIGTRAP` signal to that process. The `FiniteType` runtime class is updated to return a user-defined type name if available when the `getName()` method is called on it. Also, runtime APIs are now modified to return the type-reference type instances according to the type definitions.
+
+## Standard library updates
+
+- The Update 4 release introduces the `persist package`, which provides the Ballerina persistent layer functionality of conveniently storing and querying data in a data store. New persist CLI tooling is also introduced along with this to enable the Ballerina persistence layer in a Ballerina project to generate types and clients based on the data model. This is an experimental feature. The APIs and tooling experience might change in future releases.
+- The `graphql` package is improved such that now, a single client can have multiple subscriptions (multiplexing), and if additional information about a particular field of a GraphQL document (e.g., subfields of a field) is required, that information can be accessed inside a resolver or an interceptor method.
+- A new JetStream client is introduced to the `nats` package to carry out the publishing and subscribing streaming functionalities related to the NATS JetStreams feature.
+- The `constraint` package is enhanced by adding constraint validation support for read-only types.
+- The `oauth2` package now allows the use of inferred values for refresh token configurations through the password grant type, and string values for the `scopes` field in the client grant configuration.
+
+## Code to cloud updates
+
+Now, you can disable creating thin JAR files when generating the Docker image in order to reduce the build time of the generation. However, this will disable the Docker layering feature.  Also, the `CMD` property of the Dockerfile is improved to support complex package names.
+
+## Developer tools updates
+
+- This Update 4 release introduces a new `Create variable with Type` code action along with `Quick Pick` support for selecting expressions in the `Extract to constant` code action. Additionally, the `Rename popup` feature was added to the `Extract to constant`, `Extract to local variable`, and 'Extract to function' code actions. 
+- From this release onwards, you can generate Ballerina client and service declarations from Swagger 2.0 (i.e., OpenAPI 2.0) definitions along with a few other improvements done for OpenAPI object schema attributes and the default request and response types of the generated Ballerina resource/remote methods.
```

Comment:
Shall we split into two sentences as below, as having a single sentence might make the idea complicated IMO?
```suggestion
- From this release onwards, you can generate Ballerina client and service declarations from Swagger 2.0 (i.e., OpenAPI 2.0) definitions. Along with that, also a few other improvements were introduced which are related to OpenAPI object schema attributes and, the default request and response types of the generated Ballerina resource/remote methods.
```

Reply:
Updated.